realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyMediatorGenerator.java

```diff
@@ -0,0 +1,333 @@
+        writer.emitEmptyLine();
+    }
+
+    private void emitCopyToRealmMethod(JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod(
+                "<E extends RealmObject> E",
+                "copyOrUpdate",
+                EnumSet.of(Modifier.PUBLIC),
+                "Realm", "realm", "E", "obj", "boolean", "update", "Map<RealmObject, RealmObjectProxy>",  "cache"
+        );
+
+        writer.emitStatement("Class<E> clazz = (Class<E>) ((obj instanceof RealmObjectProxy) ? obj.getClass().getSuperclass() : obj.getClass())");
+        writer.emitEmptyLine();
+        emitMediatorSwitch(new ProxySwitchStatement() {
+            @Override
+            public void emitStatement(int i, JavaWriter writer) throws IOException {
+                writer.emitStatement("return (E) %s.copyOrUpdate(realm, (%s) obj, update, cache)", proxyClasses.get(i), simpleModelClasses.get(i));
+            }
+        }, writer, false);
+        writer.endMethod();
+        writer.emitEmptyLine();
+    }
+
+    private void emitCreteOrUpdateUsingJsonObject(JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod(
+                "<E extends RealmObject> E",
+                "createOrUpdateUsingJsonObject",
+                EnumSet.of(Modifier.PUBLIC),
+                Arrays.asList("Class<E>", "clazz", "Realm", "realm", "JSONObject", "json", "boolean", "update"),
+                Arrays.asList("JSONException")
+        );
+        emitMediatorSwitch(new ProxySwitchStatement() {
+            @Override
+            public void emitStatement(int i, JavaWriter writer) throws IOException {
+                writer.emitStatement("return (E) %s.createOrUpdateUsingJsonObject(realm, json, update)", proxyClasses.get(i));
+            }
+        }, writer);
+        writer.endMethod();
+        writer.emitEmptyLine();
+    }
+
+    private void emitCreateUsingJsonStream(JavaWriter writer) throws IOException {
+        writer.emitAnnotation("Override");
+        writer.beginMethod(
+                "<E extends RealmObject> E",
+                "createUsingJsonStream",
+                EnumSet.of(Modifier.PUBLIC),
+                Arrays.asList("Class<E>", "clazz", "Realm", "realm", "JsonReader", "reader"),
+                Arrays.asList("java.io.IOException")
+        );
+        emitMediatorSwitch(new ProxySwitchStatement() {
+            @Override
+            public void emitStatement(int i, JavaWriter writer) throws IOException {
+                writer.emitStatement("return (E) %s.createUsingJsonStream(realm, reader)", proxyClasses.get(i));
+            }
+        }, writer);
+        writer.endMethod();
+        writer.emitEmptyLine();
+    }
+
+    // Emits the control flow for selecting the appropriate proxy class based on the model class
+    // Currently it is just if..else, which is inefficient for large amounts amounts of model classes.
```

Comment:
I have been playing around with a handcoded HashMap implementation for this compared to this if/else approach, and the numbers are not that great:

Test scenario: 20 model classes, trying to call method on every class 10.000 times, Runs on GenyMotion.

```
I/Speed﹕ List: Creating class: 1.286.494 ns.
I/Speed﹕ List: All classes (10000): 18ms.
I/Speed﹕ HashMap: Creating class: 3.091.176 ns.
I/Speed﹕ HashMap: All classes (10000): 13ms.

I/Speed﹕ List: Creating class: 724.824 ns.
I/Speed﹕ List: All classes (10000): 16ms.
I/Speed﹕ HashMap: Creating class: 2.814.679 ns.
I/Speed﹕ HashMap: All classes (10000): 14ms.

I/Speed﹕ List: Creating class: 670.570 ns.
I/Speed﹕ List: All classes (10000): 15ms.
I/Speed﹕ HashMap: Creating class: 2.031.939 ns.
I/Speed﹕ HashMap: All classes (10000): 13ms.
```
Basically we are trading a bit more runtime speed for A LOT of upfront initialisation. I only converted two of 8 methods to a HashMap implementation and each one adds a lot of overhead as they need to wrap each method call in a anonymous inner class, so we can only expect initialisation to go up as well as the increased memory usage.

So I think sticking with if/else for the time being is the best way.